File: 255. Verify Preorder Sequence in Binary Search Tree.java
class Solution {
    /**
     * Using stack for mimicing the inorder traversal,
     * if we have poped item out of the stack,
     * which means that there will not be a smaller item than it
     * Time O(n) Space O(n)
     * @param preorder
     * @return
     */
    public boolean verifyPreorder(int[] preorder) {
        Deque<Integer> stack = new ArrayDeque<>();
        int low = Integer.MIN_VALUE;
        for (int i : preorder) {
            if (i < low) {
                return false;
            }
            while (!stack.isEmpty() && stack.peek() < i) {
                low = stack.pop();
            }
            stack.push(i);
        }
        return true;
    }
}

File: 255. Verify Preorder Sequence in Binary Search Tree.py
class Solution:
    def verifyPreorder(self, preorder: List[int]) -> bool:
        stack = [float("inf")]
        lo = float("-inf")
        for i in preorder:
            while i > stack[-1]:
                temp = stack.pop()
                lo = temp
            if lo >= i:
                return False
            stack.append(i)
        return True